# Exploring Toronto Neighbourhoods

### Importing Librairies

In [1]:
import numpy as np 
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 
from geopy.geocoders import Nominatim

import requests 

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans
import folium 

In [32]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [2]:
import geocoder

## Part 1 - Retrieving the data from Wikipedia

In [3]:
from bs4 import BeautifulSoup 

In [4]:
req = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M") 
soup = BeautifulSoup(req.content,'lxml') 

table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned': #This line will allow not to retrieve the postal code that are not assigned
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

print(table_contents)

[{'PostalCode': 'M3A', 'Borough': 'North York', 'Neighborhood': 'Parkwoods'}, {'PostalCode': 'M4A', 'Borough': 'North York', 'Neighborhood': 'Victoria Village'}, {'PostalCode': 'M5A', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Regent Park, Harbourfront'}, {'PostalCode': 'M6A', 'Borough': 'North York', 'Neighborhood': 'Lawrence Manor, Lawrence Heights'}, {'PostalCode': 'M7A', 'Borough': "Queen's Park", 'Neighborhood': 'Ontario Provincial Government'}, {'PostalCode': 'M9A', 'Borough': 'Etobicoke', 'Neighborhood': 'Islington Avenue'}, {'PostalCode': 'M1B', 'Borough': 'Scarborough', 'Neighborhood': 'Malvern, Rouge'}, {'PostalCode': 'M3B', 'Borough': 'North York', 'Neighborhood': 'Don Mills North'}, {'PostalCode': 'M4B', 'Borough': 'East York', 'Neighborhood': 'Parkview Hill, Woodbine Gardens'}, {'PostalCode': 'M5B', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Garden District, Ryerson'}, {'PostalCode': 'M6B', 'Borough': 'North York', 'Neighborhood': 'Glencairn'}, {'PostalCode': 'M9

In [5]:
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})


### Exploring the data

In [6]:
df.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4                       Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                     Don Mills North  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...  
12             Rouge Hill, Port Union, Highland Creek  
13                                    Don Mills South  
14                                   Woodbine Heights

In [7]:
df.describe(include = 'all')

PostalCode     Borough                       Neighborhood
count         103         103                                103
unique        103          15                                103
top           M1H  North York  Guildwood, Morningside, West Hill
freq            1          24                                  1

In [8]:
print(df.shape)

(103, 3)


Since there are 103 lines and Postal Codes and Neighbourhood name are also 103, we can conclude that we have no duplicates in our dataframe

## Part 2 - Importing Latitude and Longitude data

In [6]:
data_coordinates = pd.read_csv('Geospatial_Coordinates.csv')


In [7]:
data_coordinates.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [8]:
df.rename(columns={'PostalCode': 'Postal Code'}, inplace = True)


In [9]:
df2 = pd.merge(df,data_coordinates)

In [10]:
df2.head()

Postal Code           Borough                      Neighborhood   Latitude  \
0         M3A        North York                         Parkwoods  43.753259   
1         M4A        North York                  Victoria Village  43.725882   
2         M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3         M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4         M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [19]:
df2.shape

(103, 5)

## Part 3 - Exploring Toronto Boroughs

We'll retain only the borough containing "Toronto" in them

In [11]:
df_toronto_borough = df2[df2['Borough'].str.contains('Toronto')]

In [12]:
df_toronto_borough.reset_index(inplace=True)

In [16]:
df_toronto_borough

Postal Code                 Borough  \
0          M5A        Downtown Toronto   
1          M5B        Downtown Toronto   
2          M5C        Downtown Toronto   
3          M4E            East Toronto   
4          M5E        Downtown Toronto   
5          M5G        Downtown Toronto   
6          M6G        Downtown Toronto   
7          M5H        Downtown Toronto   
8          M6H            West Toronto   
9          M4J  East York/East Toronto   
10         M5J        Downtown Toronto   
11         M6J            West Toronto   
12         M4K            East Toronto   
13         M5K        Downtown Toronto   
14         M6K            West Toronto   
15         M4L            East Toronto   
16         M5L        Downtown Toronto   
17         M4M            East Toronto   
18         M4N         Central Toronto   
19         M5N         Central Toronto   
20         M4P         Central Toronto   
21         M5P         Central Toronto   
22         M6P            West Toronto   
23         M4R         Central Toronto   
24         M5R         Central Toronto   
25         M6R            West Toronto   
26         M4S         Central Toronto   
27         M5S        Downtown Toronto   
28         M6S            West Toronto   
29         M4T         Central Toronto   
30         M5T        Downtown Toronto   
31         M4V         Central Toronto   
32         M5V        Downtown Toronto   
33         M4W        Downtown Toronto   
34         M5W  Downtown Toronto Stn A   
35         M4X        Downtown Toronto   
36         M5X        Downtown Toronto   
37         M4Y        Downtown Toronto   
38         M7Y   East Toronto Business   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1                            Garden District, Ryerson  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
3                                         The Beaches  43.676357 -79.293031  
4                                         Berczy Park  43.644771 -79.373306  
5                                  Central Bay Street  43.657952 -79.387383  
6                                            Christie  43.669542 -79.422564  
7                            Richmond, Adelaide, King  43.650571 -79.384568  
8                        Dufferin, Dovercourt Village  43.669005 -79.442259  
9                                  The Danforth  East  43.685347 -79.338106  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21                           Forest Hill North & West  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                                 North Toronto West  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.65

### Create a map of Toronto with its neighbourhoods

In [17]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [18]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df_toronto_borough['Latitude'], df_toronto_borough['Longitude'], df_toronto_borough['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Clustering of neighbourhood, starting with 5 clusters

In [19]:
# set number of clusters
kclusters = 5

df_toronto_borough_clustering = df_toronto_borough.drop(['Neighborhood','Postal Code','Borough'], axis = 1)

# # run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0)
kmeans.fit(df_toronto_borough_clustering)
# # check cluster labels generated for each row in the dataframe
kmeans.labels_[:]

array([0, 0, 0, 3, 0, 0, 2, 0, 4, 3, 0, 2, 3, 0, 2, 3, 0, 3, 1, 1, 1, 1,
       4, 1, 2, 4, 1, 2, 4, 1, 2, 1, 0, 0, 0, 0, 0, 0, 3])

In [20]:
df_toronto_borough_clustering['Cluster Label'] = kmeans.labels_
df_toronto_borough_clustering['Neighborhood'] = df_toronto_borough['Neighborhood']

In [21]:
df_toronto_borough_clustering

Latitude  Longitude  Cluster Label  \
0   43.654260 -79.360636              0   
1   43.657162 -79.378937              0   
2   43.651494 -79.375418              0   
3   43.676357 -79.293031              3   
4   43.644771 -79.373306              0   
5   43.657952 -79.387383              0   
6   43.669542 -79.422564              2   
7   43.650571 -79.384568              0   
8   43.669005 -79.442259              4   
9   43.685347 -79.338106              3   
10  43.640816 -79.381752              0   
11  43.647927 -79.419750              2   
12  43.679557 -79.352188              3   
13  43.647177 -79.381576              0   
14  43.636847 -79.428191              2   
15  43.668999 -79.315572              3   
16  43.648198 -79.379817              0   
17  43.659526 -79.340923              3   
18  43.728020 -79.388790              1   
19  43.711695 -79.416936              1   
20  43.712751 -79.390197              1   
21  43.696948 -79.411307              1   
22  43.661608 -79.464763              4   
23  43.715383 -79.405678              1   
24  43.672710 -79.405678              2   
25  43.648960 -79.456325              4   
26  43.704324 -79.388790              1   
27  43.662696 -79.400049              2   
28  43.651571 -79.484450              4   
29  43.689574 -79.383160              1   
30  43.653206 -79.400049              2   
31  43.686412 -79.400049              1   
32  43.628947 -79.394420              0   
33  43.679563 -79.377529              0   
34  43.646435 -79.374846              0   
35  43.667967 -79.367675              0   
36  43.648429 -79.382280              0   
37  43.665860 -79.383160              0   
38  43.662744 -79.321558              3   

                                         Neighborhood  
0                           Regent Park, Harbourfront  
1                            Garden District, Ryerson  
2                                      St. James Town  
3                                         The Beaches  
4                                         Berczy Park  
5                                  Central Bay Street  
6                                            Christie  
7                            Richmond, Adelaide, King  
8                        Dufferin, Dovercourt Village  
9                                  The Danforth  East  
10  Harbourfront East, Union Station, Toronto Islands  
11                           Little Portugal, Trinity  
12                       The Danforth West, Riverdale  
13           Toronto Dominion Centre, Design Exchange  
14       Brockton, Parkdale Village, Exhibition Place  
15                     India Bazaar, The Beaches West  
16                     Commerce Court, Victoria Hotel  
17                                    Studio District  
18                                      Lawrence Park  
19                                           Roselawn  
20                                   Davisville North  
21                           Forest Hill North & West  
22                      High Park, The Junction South  
23                                 North Toronto West  
24                The Annex, North Midtown, Yorkville  
25                             Parkdale, Roncesvalles  
26                                         Davisville  
27                     University of Toronto, Harbord  
28                                 Runnymede, Swansea  
29                        Moore Park, Summerhill East  
30          Kensington Market, Chinatown, Grange Park  
31  Summerhill West, Rathnelly, South Hill, Forest...  
32  CN Tower, King and Spadina, Railway Lands, Har...  
33                                           Rosedale  
34                                     Enclave of M5E  
35                        St. James Town, Cabbagetown  
36             First Canadian Place, Underground city  
37                               Church and Wellesley  
38                                     Enclave of M4L

In [24]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(df_toronto_borough_clustering['Latitude'], df_toronto_borough_clustering['Longitude'], df_toronto_borough_clustering['Neighborhood'], df_toronto_borough_clustering['Cluster Label']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters